### PR TITLE
minogrpc: Implements message relaying

### PR DESCRIPTION
**Pull request by nkcr**
Adds a simple relaying mechanism: if a client must send a message to someone it doesn't know, it packs the message in an enveloppe and sends it to the orchestrator that relays the message. 

More tests to increase coverage will come.

### CODE REVIEW

File: mino/minogrpc/overlay.proto [hunk 1]
```diff
@@ -8,7 +8,7 @@ import "google/protobuf/any.proto";
 message Envelope {
     string from = 1;
     repeated string to = 2;
-    google.protobuf.Any message = 3;
+    google.protobuf.Any message = 4;
```
Comment (Gilthoniel) [inline mino/minogrpc/overlay.proto]: ?

File: mino/minogrpc/server.go [hunk 2]
```diff
@@ -187,7 +187,7 @@ func (rpc RPC) Stream(ctx context.Context,
 
 		header := metadata.New(map[string]string{
 			headerURIKey:     rpc.uri,
-			headerAddressKey: ""})
+			headerAddressKey: addr.String()})
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go]: Can you add a comment to explain what is this address ?

File: mino/minogrpc/server.go [hunk 3]
```diff
@@ -215,21 +215,60 @@ func (rpc RPC) Stream(ctx context.Context,
...
+					}
+				} else {
+					anyEnvelope, err := ptypes.MarshalAny(&Envelope{
+						Message: msg.Message})
+					if err != nil {
+						fillAndLog(errs, encoding.NewAnyEncodingError(anyEnvelope, err))
+						return
+					}
+
+					orchRecv.in <- &OverlayMsg{Message: anyEnvelope}
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go]: Do you really need to have both types of message ? Isn't an envelop simply an overlay message ?

File: mino/minogrpc/server.go [hunk 4]
```diff
@@ -215,21 +215,60 @@ func (rpc RPC) Stream(ctx context.Context,
...
+				}
 			}
 		}()
 	}
 
 	return orchSender, orchRecv
 }
 
+func relayMessage(enveloppe *Envelope, orchSender sender) error {
+	var dynamicAny ptypes.DynamicAny
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go]: It looks like it is not used.

File: mino/minogrpc/server.go [hunk 5]
```diff
@@ -472,17 +549,26 @@ func (r receiver) Recv(ctx context.Context) (mino.Address, proto.Message, error)
 
 	enveloppe := &Envelope{}
 	err = ptypes.UnmarshalAny(msg.Message, enveloppe)
-	if err != nil {
-		return nil, nil, encoding.NewAnyDecodingError(msg.Message, err)
+	// we can receive two types of messages: enveloppes and other messages. If
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go]: Why don't you make the logic easier by always sending an envelope/overlay message ? When a message is received, you check that the address is yours or if you should relay.

File: mino/minogrpc/server_test.go [hunk 6]
```diff
@@ -688,66 +655,171 @@ func TestRPC_MultipleChange_Stream(t *testing.T) {
...
+	dummyMsg2, ok := msg2.(*wrappers.StringValue)
+	require.True(t, ok)
+	require.Equal(t, "dummy_value_"+identifier1+"_"+identifier2+"_"+identifier3+"_"+identifier4,
+		dummyMsg2.Value)
+
+	server1.grpcSrv.GracefulStop()
+	server2.grpcSrv.GracefulStop()
+	server3.grpcSrv.GracefulStop()
+	server4.grpcSrv.GracefulStop()
+	require.NoError(t, err)
```
Comment (Gilthoniel) [inline mino/minogrpc/server_test.go]: what error ?

File: mino/minogrpc/server.go (hunk 4 above)
Comment (nkcr) [reply mino/minogrpc/server.go]: line 227 ; )

File: mino/minogrpc/server.go (hunk 2 above)
Comment (nkcr) [reply mino/minogrpc/server.go]: When a server wants to send a message to someone it doesn't know, it then creates an enveloppe with the "from" and "to" and sends it back. However, a server does not know its own address, this is why we set it in the header metadata.

If I make a better use of the enveloppe messages, I should be able to get rid of this attribute (the sever could just read the "to" attribute of the enveloppe to know its own address).

File: mino/minogrpc/server.go (hunk 3 above)
Comment (nkcr) [reply mino/minogrpc/server.go]: True, I was getting closer to this conclusion. I am just concerned that using only enveloppes will be insufficient for more advanced routing. What do you think?

File: mino/minogrpc/server_test.go (hunk 6 above)
Comment (nkcr) [reply mino/minogrpc/server_test.go]: oops, that's the rest of a non-graceful server stop

File: mino/minogrpc/server.go (hunk 2 above)
Comment (Gilthoniel) [reply mino/minogrpc/server.go]: I'm not sure to understand why a server couldn't know it's own address. Can't you instantiate the service with the address given when instantiating mino ?

Comment (nkcr) [reply mino/minogrpc/server.go]: that's definitely doable!

**Pull request merged**